### PR TITLE
Add device attestation to signup flow

### PR DESCRIPTION
**Pull request by haileyok**
This PR adds device attestation token creation (using the library found at https://github.com/bluesky-social/react-native-device-attest) for use in entryway. Entryway will render a judgement on whether a captcha is required based on the attestation results provided by the respective APIs. Attestation only happens on native clients, not on web.

## Testing

Upon an initial sign up on a device (currently only on staging, which requires an invite code found in Slack), you should not have to enter a captcha. On subsequent signups, you should be given a captcha to complete.

The main thing to pay attention to is:
- When going through an "attestation" signup, the screen shouldn't immediately flash away (minimum delay of 3.5 seconds)
- No errors should be presented if you do in fact have to enter a captcha, you should just be presented with a captcha.

### CODE REVIEW

File: src/screens/Signup/index.tsx [hunk 1]
```diff
@@ -101,6 +104,14 @@ export function Signup({onPressBack}: {onPressBack: () => void}) {
     return () => subscription.remove()
   }, [])
 
+  // On Android, warmup the Play Integrity API on the signup screen so it is ready by the time we get to the gate screen.
+  useEffect(() => {
+    if (!isAndroid) {
+      return
+    }
+    ReactNativeDeviceAttest.warumpIntegrity(GCP_PROJECT_ID)
```
Comment (surfdude29) [inline src/screens/Signup/index.tsx]: maybe a typo here, `warump` instead of `warmup`?

File: src/screens/Signup/StepCaptcha/CaptchaWebView.tsx [hunk 2]
```diff
@@ -13,6 +13,7 @@ const ALLOWED_HOSTS = [
   'js.hcaptcha.com',
   'newassets.hcaptcha.com',
   'api2.hcaptcha.com',
+  'e9755d3e012b.ngrok.app',
```
Comment (haileyok) [inline src/screens/Signup/StepCaptcha/CaptchaWebView.tsx]: TODO: remove

File: src/screens/Signup/StepCaptcha/CaptchaWebView.tsx [hunk 3]
```diff
@@ -29,12 +30,13 @@ export function CaptchaWebView({
...
-      new URL(state?.serviceUrl).host === 'staging.bsky.dev'
-      ? 'staging.bsky.app'
-      : 'bsky.app'
+    return 'e9755d3e012b.ngrok.app'
+    // if (!state?.serviceUrl) return 'bsky.app'
+    //
+    // return state?.serviceUrl &&
+    //   new URL(state?.serviceUrl).host === 'staging.bsky.dev'
+    //   ? 'staging.bsky.app'
+    //   : 'bsky.app'
```
Comment (haileyok) [inline src/screens/Signup/StepCaptcha/CaptchaWebView.tsx]: Also need to undo this

File: src/screens/Signup/StepCaptcha/index.tsx [hunk 4]
```diff
@@ -1,21 +1,72 @@
...
+import {isAndroid, isIOS, isNative, isWeb} from '#/platform/detection'
 import {ScreenTransition} from '#/screens/Login/ScreenTransition'
 import {useSignupContext} from '#/screens/Signup/state'
 import {CaptchaWebView} from '#/screens/Signup/StepCaptcha/CaptchaWebView'
 import {atoms as a, useTheme} from '#/alf'
 import {FormError} from '#/components/forms/FormError'
 import {BackNextButtons} from '../BackNextButtons'
 
-const CAPTCHA_PATH = '/gate/signup'
+const CAPTCHA_PATH = isWeb ? '/gate/signup' : '/gate/signup/attempt-attest'
```
Comment (haileyok) [inline src/screens/Signup/StepCaptcha/index.tsx]: `attempt-attest` will redirect back to `/gate/signup` if attestation fails or is for some reason not availalbe

File: src/screens/Signup/index.tsx [hunk 5]
```diff
@@ -101,6 +105,16 @@ export function Signup({onPressBack}: {onPressBack: () => void}) {
...
+  // On Android, warmup the Play Integrity API on the signup screen so it is ready by the time we get to the gate screen.
+  useEffect(() => {
+    if (!isAndroid) {
+      return
+    }
+    ReactNativeDeviceAttest.warmupIntegrity(GCP_PROJECT_ID).catch(err =>
+      logger.error(err),
+    )
+  }, [])
+
```
Comment (haileyok) [inline src/screens/Signup/index.tsx:117]: The generated items - an `Integrity` provider and an "appset id" will be held on the native side. We just need to warm them up (these can take a matter of 5-10 seconds, so we get them done asap)

File: src/screens/Signup/index.tsx (hunk 1 above)
Comment (surfdude29) [reply src/screens/Signup/index.tsx]: @haileyok might need tweaking here too? 👀

https://github.com/bluesky-social/react-native-device-attest/blob/850f4e29058d33121725146a6d1ba45e2a379143/src/DeviceAttestBase.ts#L8

https://github.com/bluesky-social/react-native-device-attest/blob/850f4e29058d33121725146a6d1ba45e2a379143/src/ReactNativeDeviceAttest.android.ts#L7-L8

Comment (haileyok) [reply src/screens/Signup/index.tsx]: its fixed in https://github.com/bluesky-social/react-native-device-attest/pull/1

Comment (surfdude29) [reply src/screens/Signup/index.tsx]: oops, missed that branch, sorry

**Review by mozzius — APPROVED** [review]
I successfully did the flow - looked as it should but pretty hard to test edge cases since you can only do it once!

File: src/screens/Signup/StepCaptcha/CaptchaWebView.tsx [hunk 6]
```diff
@@ -28,49 +30,67 @@ export function CaptchaWebView({
   onSuccess: (code: string) => void
   onError: (error: unknown) => void
 }) {
-  const redirectHost = React.useMemo(() => {
+  const startedAt = useRef(Date.now())
+  const successTo = useRef<NodeJS.Timeout>()
+
+  useEffect(() => {
+    return () => {
+      if (successTo) {
```
Comment (estrattonbailey) [inline src/screens/Signup/StepCaptcha/CaptchaWebView.tsx]:
```suggestion
      if (successTo.current) {
```

Just double checking, need this?

Comment (haileyok) [reply src/screens/Signup/StepCaptcha/CaptchaWebView.tsx]: oop yes, thank you

**Review by estrattonbailey — APPROVED** [review]
Already stamped, but looks good. Will wait to test in TF later today 👍

**Comment by haileyok** [comment]
can merge once entryway pr goes up, would be good to get a full, clean build in there to test from.
